src/main/java/bio/terra/workspace/db/DataReferenceDao.java

```diff
@@ -0,0 +1,142 @@
+              rs.getString("resource_id") == null
+                  ? null
+                  : resourceDescriptionMapper.mapRow(rs, rowNum))
+          .credentialId(rs.getString("credential_id"))
+          .cloningInstructions(
+              CloningInstructionsEnum.fromValue(rs.getString("cloning_instructions")))
+          .referenceType(ReferenceTypeEnum.fromValue(rs.getString("reference_type")))
+          .reference(rs.getString("reference"));
+    }
+  }
+
+  // Returns a SQL condition as a string that selects controlled resources, uncontrolled resources,
+  // or both, depending on the value of filterControlled.
+  public static String filterControlledWhereClause(
+      String filterControlled, String referenceTableAlias) {
+    // filterControlled is one of "controlled", "uncontrolled", or "all" (enum in OpenAPI yaml).
+    // This corresponds to the workspace_data_reference.resource_id column being not null, null, or
+    // either, respectively.
+    if (filterControlled.equals("controlled")) {
+      return "(" + referenceTableAlias + ".resource_id IS NOT NULL)";
+    } else if (filterControlled.equals("uncontrolled")) {
+      return "(" + referenceTableAlias + ".resource_id IS NULL)";
+    } else if (filterControlled.equals("all")) {
+      return "";
+    } else {
+      // This case should not happen, as OpenAPI should enforce that this value is an enum.
+      throw new ValidationException("Invalid value for filterControlled: " + filterControlled);
+    }
+  }
+
+  // Returns a SQL condition as a string accepts both uncontrolled data references and visible
+  // controlled references. Uncontrolled references are not tracked as resources, and their
+  // existence is always visible to all workspace readers.
+  public static String uncontrolledOrVisibleResourcesClause(
+      String owner, String resourceTableAlias, String referenceTableAlias) {
+    return "(("
+        + referenceTableAlias
+        + ".resource_id IS NULL) OR "
+        + visibleResourcesClause(owner, resourceTableAlias)
+        + ")";
+  }
+
+  // Returns a SQL condition as a string that filters out invisible controlled references.
+  // References are considered 'invisible' if the is_visible column of the corresponding resource
+  // is false AND the resource owner is not the user issuing the query.
+  // Uncontrolled references are not tracked as resources, and their existence is always visible
+  // to all workspace readers.
+  public static String visibleResourcesClause(String owner, String resourceTableAlias) {
+    return "("
+        + resourceTableAlias
+        + ".is_visible = true OR "
+        + resourceTableAlias
+        + ".owner = \'"
+        + owner
+        + "\')";
+  }
+
+  // Combines a list of String SQL conditions with the delimiter `" AND "` to create a single
+  // SQL `WHERE` clause. Ignores null and empty strings.
+  public static String combineWhereClauses(List<String> clauses) {
+    return "WHERE ("
```

Comment:
nit: nothing wrong with it, but you don't need the parens.

Reply:
Agreed it's unnecessary, but it was a little easier for me to parse this way when I was debugging. If that's weird style for SQL I can take them out